### PR TITLE
Optimize GPT2

### PR DESCRIPTION
**Pull request by feihugis**
This PR optimizes the duplicated cache for GPT2.

### CODE REVIEW

**Comment by JiushengChen** [comment]
Please add benchmark and update readmes.

**Comment by feihugis** [comment]
> Please add benchmark and update readmes.

@JiushengChen The current benchmarking script is not compatible with GPT2 model, such as [here](https://github.com/microsoft/fastseq/blob/main/fastseq_cli/transformers_generate.py#L154): gpt2 uses `AutoModelForCausalLM` instead of `AutoModelForSeq2SeqLM`. I did some hardcode changes locally to run the benchmarks. Will fix the compatibility issue in the next commit.

**Comment by feihugis** [comment]
The initial benchmarking result is 26.3 samples/s (fastseq) vs 7.4 samples/s (huggingface).

**Comment by JiushengChen** [comment]
> The initial benchmarking result is 26.3 samples/s (fastseq) vs 7.4 samples/s (huggingface).

Looks good!

**Comment by feihugis** [comment]
The benchmarking results:

| Util | Model | Task | Split | BatchSize | Samples | Tokens | Bleu | Rouge | Loss | Perplexity | Runtime(seconds) | Throughput(samples/s) | Throughput(tokens/s) |
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 16 | 59815 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 5336 | 11.2 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 32 | 59864 | NA | NA | 0.050\|0.002\|0.036 | NA | NA | 4011 | 14.9 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 64 | 60283 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 3793 | 15.9 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 16 | 59815 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 5127 | 11.7 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 32 | 59864 | NA | NA | 0.050\|0.002\|0.036 | NA | NA | 4033 | 14.8 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 64 | 60283 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 3783 | 15.9 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 16 | 59815 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 5109 | 11.7 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 32 | 59864 | NA | NA | 0.050\|0.002\|0.036 | NA | NA | 4066 | 14.7 | NA |
| transformers_v3.0.2 | gpt2 | cnn_dm/raw | val | 64 | 60283 | NA | NA | 0.049\|0.002\|0.035 | NA | NA | 3816 | 15.8 | NA |
| Util | Model | Task | Split | BatchSize | Samples | Tokens | Bleu | Rouge | Loss | Perplexity | Runtime(seconds) | Throughput(samples/s) | Throughput(tokens/s) |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 16 | 66550 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 3246 | 20.5 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 32 | 66004 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 1712 | 38.6 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 64 | 66574 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 968 | 68.8 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 128 | 66101 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 609 | 108.5 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 16 | 66550 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 3372 | 19.7 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 32 | 66004 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 1744 | 37.8 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 64 | 66574 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 1011 | 65.8 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 128 | 66101 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 618 | 107.0 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 16 | 66550 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 3377 | 19.7 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 32 | 66004 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 1753 | 37.7 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 64 | 66574 | NA | NA | 0.053\|0.002\|0.039 | NA | NA | 956 | 69.6 | NA |
| transformers_v3.0.2+fastseq_v0.0.4 | gpt2 | cnn_dm/raw | val | 128 | 66101 | NA | NA | 0.054\|0.002\|0.039 | NA | NA | 626 | 105.6 | NA |

**Comment by JiushengChen** [comment]
> 66550

Samples are different. Looks some batches were dropped.

File: benchmarks/benchmark_hf.sh [hunk 1]
```diff
@@ -22,7 +22,7 @@ if [[ $framework == transformers ]]; then
     fi
     ver=`pip show transformers | awk  '{if($1=="Version:")print $2}'`
     framework_versioned="transformers_v$ver"
-    util="python $BASELINE_REPO/examples/seq2seq/run_eval.py"
+    util="python run_eval_hf.py"
```
Comment (JiushengChen) [inline benchmarks/benchmark_hf.sh:25]: why change here? Looks we don't need to copy a file and add to FastSeq git repo, Transformers is cloned and saved to local, so we can use the file there.

File: benchmarks/models/hf_gpt2.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#   <batch-sizes>
+source hf.sh
+
+# MODEL - bart large cnn from transformer
+# TASK - cnn dm val full set
+
+./benchmark.sh \
+    transformers+fastseq \
+    gpt2 \
+    cnn_dm.1k/raw \
```
Comment (JiushengChen) [inline benchmarks/models/hf_gpt2.sh:17]: one is 1k set, the other is full set.

File: benchmarks/models/hf_gpt2.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    gpt2 \
+    cnn_dm/raw \
+    val \
+    16/32/64 \
+    --task summarization \
+    --no_repeat_ngram_size 3 \
+    --max_tokenizer_length 512 \
+    --max_gen_length 711
+
+
```
Comment (JiushengChen) [inline benchmarks/models/hf_gpt2.sh:36]: Add rouge/bleu score and speed check

File: benchmarks/benchmark_hf.sh (hunk 1 above)
Comment (feihugis) [reply benchmarks/benchmark_hf.sh:25]: `run_eval.py` in huggingface-transformers repo does not support gpt2 model. This file adds some changes: 1) line 73-79: use `AutoModelForCausalLM` to load gpt2 model and specify the pad token; 2) line 93, 103, 135-140, 156-157 are added to support max_tokenizer_length and max_gen_length parameters.

File: benchmarks/models/hf_gpt2.sh (hunk 2 above)
Comment (feihugis) [reply benchmarks/models/hf_gpt2.sh:17]: Oh, this is for debug, will change it back.

File: benchmarks/models/hf_gpt2.sh (hunk 3 above)
Comment (feihugis) [reply benchmarks/models/hf_gpt2.sh:36]: Will add it once solving the pending issues.

**Comment by feihugis** [comment]
> > 66550
> 
> Samples are different. Looks some batches were dropped.

Yes, I'm debugging this issue.

**Review by JiushengChen — APPROVED** [review]

**Pull request merged**